doc/pserver.md

```diff
@@ -0,0 +1,226 @@
+        pass
+    
+    # embedding param is handled lazily   
+    def pull_embedding_param(self, reques, _):
+        pass
+
+    def get_param(self):
+        pass
+
+    def set_param(self):
+        pass
+```
+
+
+The interface of Gradient Queue could be like this:
+
+```python
+class GradientQueue(elasticdl_pb2_grpc.GradientQueueServicer):
+    def __init__(self):
+        self.grad_queue = queue.Queue()
+        
+    def push_gradient(self, request, _):
+        pass
+    
+    def get_gradient(self):
+        pass
+        
+    def put_gradient(self):
+        pass
+```
+
+### Parameters Transfromation
+
+Parameters will be transformed into different data types in different stages.
+
+Common parameter:
+
+- worker forward/backward computation: TensorFlow Variable
+- worker pushing parameter: serialized protobuf
+- KVStore parameter DB: TensorFlow Variable
+- optimizer: TensorFlow Variable
+
+
+Embedding parameter:
+
+- worker forward/backward computation: numpy ndarry based Python struct
+- worker pushing parameter: serialized protobuf
+- KVStore embedding parameter DB: numpy ndarry based Python struct
+- optimizer: TensorFlow Variable
+
+We have to define a `Tensor` proto message and a corresponding `Tensor` Python class to support all these tranformations.
+
+
+The `Tensor` proto message could be like this:
+
+```proto
+message Tensor {
+    enum DataType {
+        BOOL = 0;
+        INT16 = 1;
+        INT32 = 2;
+        INT64 = 3;
+        FP16 = 4;
+        FP32 = 5;
+        FP64 = 6;
+    }
+    string name = 1;
+    DataType data_type = 2;
+    repeated int64 dims = 3;
+    bytes content = 4;
+    repeated int64 indices = 5;
+}
+```
+
+The `Tensor` Python class could be like this:
+
+```python
+class Tensor(object):
+    def __init__(self, name=None, value=None, indices=None):
+        self.name = name
+        self.value = value
+        self.indices = indices
+```
+
+There are also some helper functions:
+
+```python
+def serialize_to_pb(tensor, pb):
+    pass
+
+def deserialize_from_pb(pb, tensor):
+    pass
+
+def convert_to_tf_variable(tensor):
+    pass
+
+def convert_to_tf_tensor(tensor):
+    pass
+```
+
+PServer will store a subset of the full model. And a worker will push/pull a submodel from the pserver. The model message is defined as following:
+
+```proto
+message Model {
+    int64 version = 1;
+    repeated Tensor tensors = 2;
+}
+```
+
+Model could also be used as gradients collection.
+
+So the RPC service will be defined as following:
+
+```proto
+message EmbeddingResponse{
+    Tensor value = 1;
+    repeated int64 unknown_indices = 2;
+}
+
+service KVStore{
+    rpc push_model(Model) returns (google.protobuf.Empty) {}
+    rpc pull_model(Model) returns (google.protobuf.Empty) {}
+    rpc pull_embedding_param(Tensor) returns (EmbeddingResponse) {}
+}
+
+service GradientQueue {
+    rpc push_gradient(Model) returns (google.protobuf.Empty) {}
+}
+```
+
+### Gradient Queue
```

Comment:
As discussed, we will use a queue in sync-SGD. In async-SGD, we do not use the queue but apply_gradient directly in the 1st version. 
We may consider queue for async-SGD later.

Reply:
Done